show/main.py

```diff
@@ -1800,23 +1800,76 @@ def mmu():
 
 
 #
-# 'reboot-cause' command ("show reboot-cause")
-#
-@cli.command('reboot-cause')
-def reboot_cause():
-    """Show cause of most recent reboot"""
-    PREVIOUS_REBOOT_CAUSE_FILE = "/host/reboot-cause/previous-reboot-cause.txt"
-
-    # At boot time, PREVIOUS_REBOOT_CAUSE_FILE is generated based on
-    # the contents of the 'reboot cause' file as it was left when the device
-    # went down for reboot. This file should always be created at boot,
-    # but check first just in case it's not present.
-    if not os.path.isfile(PREVIOUS_REBOOT_CAUSE_FILE):
-        click.echo("Unable to determine cause of previous reboot\n")
-    else:
-        cmd = "cat {}".format(PREVIOUS_REBOOT_CAUSE_FILE)
-        proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
-        click.echo(proc.stdout.read())
+# 'reboot-cause' group ("show reboot-cause")
+#
+@cli.group('reboot-cause', invoke_without_command=True)
+@click.pass_context
+def reboot_cause(ctx):
+    if ctx.invoked_subcommand is None:
+        """Show cause of reboot"""
+        REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+        PREVIOUS_REBOOT_CAUSE_FILE = os.path.join(REBOOT_CAUSE_DIR, "previous-reboot-cause.json")
+        USER_ISSUED_REBOOT_CAUSE_REGEX ="User issued \'{}\' command [User: {}, Time: {}]"
+        REBOOT_CAUSE_UNKNOWN = "Unknown"
+
+        last_reboot_cause = REBOOT_CAUSE_UNKNOWN
+        # Read the last previous reboot cause
+        if os.path.exists(PREVIOUS_REBOOT_CAUSE_FILE):
+            with open(PREVIOUS_REBOOT_CAUSE_FILE, "r") as last_cause_file:
+                data = json.load(last_cause_file)
+                if data['user']:
+                    last_reboot_cause = USER_ISSUED_REBOOT_CAUSE_REGEX.format(data['cause'], data['user'], data['time'])
+                else:
+                    last_reboot_cause = "{}".format(data['cause'])
+
+        click.echo(last_reboot_cause)
+
+# 'history' subcommand ("show reboot-cause history")
+@reboot_cause.command()
+def history():
+    """Show history of reboot-cause"""
+    REBOOT_CAUSE_TABLE_NAME = "REBOOT_CAUSE"
+    TABLE_NAME_SEPARATOR = '|'
+    db = SonicV2Connector(host='127.0.0.1')
+    db.connect(db.STATE_DB, False)   # Make one attempt only
+    prefix = REBOOT_CAUSE_TABLE_NAME + TABLE_NAME_SEPARATOR
+    _hash = '{}{}'.format(prefix, '*')
+    table_keys = db.keys(db.STATE_DB, _hash)
+    table_keys.sort(reverse=True)
+
+    def remove_prefix(text, prefix):
+        if text.startswith(prefix):
+            return text[len(prefix):]
+        return text
+
+    table = []
+    for tk in table_keys:
+        entry = db.get_all(db.STATE_DB, tk)
+        r = []
+        r.append(remove_prefix(tk, prefix))
+        if 'cause' not in entry:
+            r.append("")
```

Comment:
Should we append the empty string, or "N/A"? Same question for all below.

Reply:
it's already add "N/A" before storing into state-db. so it should return something valid otherwise, there must be something wrong while it's saved into state-db.

Reply:
So this is just a failsafe in case something went wrong saving to State DB?

Reply:
Yes.

Reply:
If you feel it's necessary, then maybe we can just shorten the code by doing the following:

```python
r.append(entry['cause'] if 'cause' in entry else "")
r.append(entry['time'] if 'time' in entry else "")
r.append(entry['user'] if 'user' in entry else "")
r.append(entry['comment'] if 'comment' in entry else "")
```